fix(matrices): Deprecate non-Expr in Matrix

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

The deprecation issue for this is #21497
The deprecation issue for the change to `laplace_transform` is https://github.com/sympy/sympy/issues/21504

Follows on from #21441, #21427, #21402

#### Brief description of what is fixed or changed

This PR deprecates using non-Expr objects as the elements of a Matrix. The codebase previously contained a number of Matrix subclasses (`NewMatrix`, `RawMatrix` and `PolyMatrix`) that would use non-Expr elements for the entries. In most cases this was done so that the elements could be domain elements or Poly rather than Expr but a better solution to this is to use the new `DomainMatrix` class. After various other PRs it is now possible to deprecate having non-Expr elements in a Matrix.

One notable obstacle to refusing non-Expr elements is the behaviour of the `laplace_transform` function:
```python
In [2]: laplace_transform(eye(2), t, z)
Out[2]: 
⎡⎛1         ⎞                ⎤
⎢⎜─, 0, True⎟   (0, -∞, True)⎥
⎢⎝z         ⎠                ⎥
⎢                            ⎥
⎢               ⎛1         ⎞ ⎥
⎢(0, -∞, True)  ⎜─, 0, True⎟ ⎥
⎣               ⎝z         ⎠ ⎦
```
This has returned a Matrix where each element is a tuple consisting of an Expr and the convergence conditions.In this PR the above Matrix of tuples is deprecated and it is possible to obtain a new behaviour:
```python
In [2]: laplace_transform(eye(2), t, z, legacy_matrix=False)
Out[2]: 
⎛⎡1   ⎤         ⎞
⎜⎢─  0⎥         ⎟
⎜⎢z   ⎥         ⎟
⎜⎢    ⎥, 0, True⎟
⎜⎢   1⎥         ⎟
⎜⎢0  ─⎥         ⎟
⎝⎣   z⎦         ⎠
```
This returns a tuple of the transformed matrix and the convergence conditions for the matrix as a whole and is more consistent with the general behaviour of the `laplace_transform` function for other kinds of inputs. A future PR will make this behaviour the default.

#### Other comments

I plan to use `DomainMatrix` as the internal implementation of `Matrix` but that means that `Matrix` needs to be stricter about the types of its elements. Needing to accommodate for non-Expr elements in the internal routines of `Matrix` has constrained improvements to this important class for some time.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* matrices
    * DEPRECATION: Using non-Expr elements in a Matrix is now deprecated.
* integrals
    * DEPRECATION: Using `laplace_transform` with a `Matrix` and without passing `noconds=True` is now deprecated. The behaviour of this function will change in future to return the conditions separately from the Matrix. The new behaviour can be obtained by passing the keyword argument `legacy_matrix=False`. 
<!-- END RELEASE NOTES -->

smichr commented on sympy/matrices/tests/test_matrices.py:
```diff
@@ -656,9 +656,11 @@ def test_creation():
         Matrix((1, 2))[3] = 5
 
     assert Matrix() == Matrix([]) == Matrix([[]]) == Matrix(0, 0, [])
-    # anything can go into a matrix (laplace_transform uses tuples)
```
Does `laplace_transform` still do this?

smichr commented on sympy/matrices/dense.py:
```diff
@@ -313,10 +315,21 @@ def _new(cls, *args, copy=True, **kwargs):
+            if not all(issubclass(typ, Expr) for typ in types):
+                SymPyDeprecationWarning(
+                    feature="non-Expr objects in a Matrix",
+                    useinstead="list of lists or some other data structure",
```
perhaps we could advertise `TableForm` here; and maybe slicing of TableForm could be added.

oscarbenjamin replied:
That's a good point. I've mentioned TableForm in the message

oscarbenjamin committed:
fix(integrals): split out conditions in laplace_transform(Matrix)

Previously laplace_transform(Matrix) would return a Matrix where each
element is a tuple of the expression for the Laplace transform and the
convergence conditions. This commit extracts those conditions to what is
returned is the transformed matrix with all elements as expressions and
the conditions are returned separately.

oscarbenjamin committed:
fix(matrices): deprecate non-Expr elements in a Matrix.

Having non-Expr elements in a Matrix is problematic and limits
improvements to the Matrix class. This commit deprecates having non-Expr
elements in a Matrix. Future changes might make this a hard error.

oscarbenjamin committed:
doc(polys): mention TableForm in DeprecationWarning

smichr commented on sympy/integrals/transforms.py:
```diff
@@ -1181,8 +1182,13 @@ def laplace_transform(f, t, s, **hints):
+        elements_trans = [laplace_transform(fij, t, s, **hints) for fij in f]
+        elements, avals, conditions = zip(*elements_trans)
+        f_laplace = type(f)(*f.shape, elements)
+        return f_laplace, Max(*avals), And(*conditions)
```
This can't be done until the deprecation is complete, can it?

oscarbenjamin replied:
This is just fixing something that is completely broken. On master we have:
```python
In [1]: M = Matrix([[t, 0], [0, t]])

In [2]: t, s = symbols('t, s')

In [3]: M = Matrix([[t, 0], [0, t]])

In [4]: laplace_transform(M, t, s)
Out[4]: 
⎡⎛1          ⎞               ⎤
⎢⎜──, 0, True⎟  (0, -∞, True)⎥
⎢⎜ 2         ⎟               ⎥
⎢⎝s          ⎠               ⎥
⎢                            ⎥
⎢               ⎛1          ⎞⎥
⎢(0, -∞, True)  ⎜──, 0, True⎟⎥
⎢               ⎜ 2         ⎟⎥
⎣               ⎝s          ⎠⎦
```
That's a Matrix but with each element being a tuple.

The PR gives:
```python
In [3]: laplace_transform(M, t, s)
Out[3]: 
⎛⎡1     ⎤         ⎞
⎜⎢──  0 ⎥         ⎟
⎜⎢ 2    ⎥         ⎟
⎜⎢s     ⎥         ⎟
⎜⎢      ⎥, 0, True⎟
⎜⎢    1 ⎥         ⎟
⎜⎢0   ──⎥         ⎟
⎜⎢     2⎥         ⎟
⎝⎣    s ⎦         ⎠
```
IOW the PR returns the matrix plus the conditions. On master you get a matrix of tuples with the conditions as part of the tuples in the elements of the matrix. That's so broken that I don't think any deprecation is needed.

smichr replied:
It's not broken for the person that knows this and looks at a given matrix element's tuple to get the condition(s). Not that I don't want to see the change -- I think it makes much more sense like this -- but this is backwards incompatible for anyone that knows and handles this already. So should the cycle be 1) we need a keyword that allows the new output but default to the old. Then when the deprecation occurs, 2) the keyword has to default to new style and be put into deprecation, and finally 3) the keyword gets dropped.

oscarbenjamin replied:
I'm not sure this needs a deprecation cycle but I've added it in.

I realised when looking at it that actually on master reasonable results can be obtained when using `noconds=False` but this PR broke that behaviour. I've fixed that and added tests for it.

oscarbenjamin committed:
fix(integrals): Add legacy_matrix option to laplace_transform

oscarbenjamin commented:
Thanks!

I've update the OP here and also for the related deprecation issues.
